compparallel.go

```diff
@@ -0,0 +1,280 @@
+package routinghelpers
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/routing"
+)
+
+var _ routing.Routing = &Parallel{}
+
+type ComposableParallel struct {
+	routers []*ParallelRouter
+}
+
+func NewComposableParallel(routers []*ParallelRouter) *ComposableParallel {
+	return &ComposableParallel{
+		routers: routers,
+	}
+}
+
+func (r *ComposableParallel) Provide(ctx context.Context, cid cid.Cid, provide bool) error {
+	var wg sync.WaitGroup
+	errCh := make(chan error)
+	for _, r := range r.routers {
+		r := r
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			tim := time.NewTimer(r.ExecuteAfter)
+			defer tim.Stop()
```

Comment:
Slightly related: I tried to make a `forEachParallel` fn w/ a similar motivation, I havn't actually tested this yet but just throwing it out there as a starting point

```go
// forEachParallel executes the function for each router, in parallel.
// If any function returns true, then the context passed into all functions will be immediately canceled.
func (r routers) forEachParallel(ctx context.Context, f func(ctx context.Context, router routing.Routing) (cancelOthers bool, err error)) error {
	var wg sync.WaitGroup
	errCh := make(chan error)
	wg.Add(len(r))
	ctx, cancelAll := context.WithCancel(ctx)
	for _, router := range r {
		go func(pr *ParallelRouter) {
			defer wg.Done()
			var cancelOthers bool
			err := executeAfter(ctx, pr.ExecuteAfter, func() error {
				ctx, cancel := context.WithTimeout(ctx, pr.Timeout)
				defer cancel()
				co, err := f(ctx, pr.Router)
				cancelOthers = co
				return err
			})
			if cancelOthers {
				cancelAll()
			}
			if err != nil &&
				!errors.Is(err, routing.ErrNotSupported) &&
				!pr.IgnoreError {
				errCh <- err
			}
		}(router)
	}
	go func() {
		wg.Wait()
		close(errCh)
	}()

	var errOut error
	for err := range errCh {
		errOut = multierror.Append(errOut, err)
	}

	return errOut
}


```

Reply:
You need to close the `addrChanOut` after finishing the iteration. we can add a new function that will be called after all routers are executed, but it is worth it? In my opinion, we are trying to extract to a function some code that is almost the same, but it is slightly different (a lot of bootstrap/scaffolding code needed in go).

Let me think about how to reuse some parts without using `interface{}` and casts everywhere.

Reply:
WDYT about using generics? I was experimenting a bit:
```go
func getValueOrError[T any](
	ctx context.Context,
	routers []*ParallelRouter,
	f func(context.Context, routing.Routing) (T, error),
	isEmpty func(T) bool,
) (value T, err error) {
	outCh := make(chan T)
	errCh := make(chan error)

	// global cancel context to stop early other router's execution.
	ctx, gcancel := context.WithCancel(ctx)
	var wg sync.WaitGroup
	for _, r := range routers {
		wg.Add(1)
		go func(r *ParallelRouter) {
			defer wg.Done()
			tim := time.NewTimer(r.ExecuteAfter)
			defer tim.Stop()
			select {
			case <-ctx.Done():
				if !r.IgnoreError {
					errCh <- ctx.Err()
				}
			case <-tim.C:
				ctx, cancel := context.WithTimeout(ctx, r.Timeout)
				defer cancel()
				value, err := f(ctx, r.Router)
				if err != nil &&
					!errors.Is(err, routing.ErrNotFound) &&
					!r.IgnoreError {
					select {
					case <-ctx.Done():
					case errCh <- err:
					}
					return
				}
				if isEmpty(value) {
					return
				}
				select {
				case <-ctx.Done():
					return
				case outCh <- value:
				}
			}
		}(r)
	}

	// goroutine closing everything when finishing execution
	go func() {
		wg.Wait()
		close(outCh)
		close(errCh)
	}()

	select {
	case out, ok := <-outCh:
		gcancel()
		if !ok {
			return value, routing.ErrNotFound
		}
		return out, nil
	case err, ok := <-errCh:
		gcancel()
		if !ok {
			return value, routing.ErrNotFound
		}
		return value, err
	case <-ctx.Done():
		gcancel()
		return value, ctx.Err()
	}
}
````
```go
func (r *ComposableParallel) GetValue(ctx context.Context, key string, opts ...routing.Option) ([]byte, error) {
	return getValueOrError(ctx, r.routers,
		func(ctx context.Context, r routing.Routing) ([]byte, error) {
			return r.GetValue(ctx, key, opts...)
		},
		func(ai []byte) bool {
			return len(ai) == 0
		})
}
```
```go
func (r *ComposableParallel) FindPeer(ctx context.Context, id peer.ID) (peer.AddrInfo, error) {
	return getValueOrError(ctx, r.routers,
		func(ctx context.Context, r routing.Routing) (peer.AddrInfo, error) {
			return r.FindPeer(ctx, id)
		},
		func(ai peer.AddrInfo) bool {
			return ai.ID == ""
		})
}
```
I prefer to use generics as less as possible, but I think it fits well in this case.

Reply:
lgtm

Reply:
Might be simpler to just return (val, bool, err) instead of having the extra callback, like

```go
func (r *ComposableParallel) GetValue(ctx context.Context, key string, opts ...routing.Option) ([]byte, error) {
	return getValueOrError(ctx, r.routers,
		func(ctx context.Context, r routing.Routing) ([]byte, bool, error) {
			b, err := r.GetValue(ctx, key, opts...)
			return b, len(b) == 0, err
		})
}
```